File: 20 - Valid Parentheses/Java/solution.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> parens = Map.of(
            ')', '(',
            '}', '{',
            ']', '['
        );

        String open_parens = "([{";
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (open_parens.contains(String.valueOf(c))) {
                stack.push(c);
            }
            else {
                if(stack.size() == 0) {
                    return false;
                }
                
                char paren_to_check = stack.pop();
                char close_paren = parens.get(c);

                if (close_paren != paren_to_check) {
                    return false;
                }
            }
        }
        if(stack.size() == 0) {
            return true;
        }
        return false;
    }
}

File: 20 - Valid Parentheses/Python/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        open_paren = "([{"
        parens = {")": "(", "}": "{", "]": "["}
        for p in s:
            # open paren, just add to stack
            if p in open_paren:
                stack.append(p)
            else:
                # pop top of stack
                if len(stack) == 0:
                    return False
                open_to_check = stack.pop()
                if parens[p] != open_to_check:
                    return False
        
        if len(stack) == 0:
            return True
        return False
